gph/python/ripser_interface.py

```diff
@@ -558,30 +556,27 @@ def ripser_parallel(X, maxdim=1, thresh=np.inf, coeff=2, metric="euclidean",
         if weights is not None:
             dm = _compute_weights(dm, weights, weight_params, n_points)
 
-        compute_enclosing_radius = False
-        nonzero_in_diag = (dm.diagonal() != 0).any()
-        if not nonzero_in_diag:
+        apply_user_threshold = thresh != np.inf
+        if not apply_user_threshold:
             # Compute ideal threshold only when a distance matrix is passed
             # as input without specifying any threshold
             # We check if any element and if no entries is present in the
             # diagonal. This allows to have the enclosing radius before
             # calling collapser if computed
-            if thresh == np.inf:
-                thresh = _ideal_thresh(dm, thresh)
-                compute_enclosing_radius = True
-
-        if nonzero_in_diag:
-            # Convert to sparse format, because currently that's the only one
-            # handling nonzero births
-            (row, col) = np.triu_indices_from(dm)
-            data = dm[(row, col)]
-            if collapse_edges:
+            thresh = _ideal_thresh(dm, thresh)
+
+        if collapse_edges:
+            if (dm.diagonal() != 0).any():
```

Comment:
Can we not simplify (remove?) this logic now that we do support non-zero births?  At least it seems to me that the comment is now outdated.

Reply:
To be more precise/constructive here, my point is that it may be cleaner to create a `_collapse_dense` wrapper function more or less as follows:

```
def _collapse_dense(dm, thresh):
    # Diagonal data (in case it is nonzero)
    data_diag = dm.diagonal()
    row_diag = np.arange(n_points)
    col_diag = row_diag.copy()
    # Off-diagonal collapsed edges
    row, col, data = gph_collapser.\
        flag_complex_collapse_edges_dense(dm, thresh)

    return (np.hstack([row_diag, row]),
            np.hstack([col_diag, col]),
            np.hstack([data_diag, data]))
```
Then one could call ``row, col, data = _collapse_dense(dm.astype(np.float32), thresh)`` without conditionals.

Btw, why don't we convert to `np.float32` in the case of `_collapsed_coo`?

Reply:
Umbe, you discovered a bug ...
In collapser, the diagonal was always discarded, dense and sparse case.
I will push a commit here that you can have a look and let me know what do you think.

The commit fixes the bindings to include the diagonal. In the if condition for dense and sparse case, you will see 2 changes:
1. Include diagonal
2. Check if the value in the entry is different than 0. (To prevent including 0 entries from a 0-diagonal per example)

I also made slightly different function as you suggested, let me know what do you think.

Reply:
> Btw, why don't we convert to np.float32 in the case of _collapse_coo?

I think I was just inconsistent ...

Reply:
Wait, there was no bug, it was by purpose that we did not include the diagonal in collapser.
Sorry for that, but anyway, I will revert just the part related to collapser bindings.

Reply:
Thanks @MonkeyBreaker!  I agree that there was no bug.

> I also made slightly different function as you suggested, let me know what do you think.

Thanks!  I still think that this logic deserves to be encapsulated into a function as I initially suggested.  After all, we do something logically very similar to `_collapse_coo`, so it would be more consistent.  Furthermore, I still slightly favour the original code I suggested over the use of `diags`, because using `diags` is slower (about 300 times slower in my machine!) as it performs several checks and format conversions internally.  It's true that using `diags` lets us skip some nonzero values if there are any, but the memory impact of storing the entire diagonal is negligible anyway.

> I think I was just inconsistent ...

Which way should we go to be consistent?  Since the collapser bindings use 64-bit, I am tempted to say that we should probably delay conversion to 32-bit to later, so we can use the extra precision if necessary until we really *have* to downcast inside ripser.

Reply:
> Thanks! I still think that this logic deserves to be encapsulated into a function as I initially suggested. After all, we do something logically very similar to _collapse_coo, so it would be more consistent.

Sure !

>  still slightly favour the original code I suggested over the use of diags, because using diags is slower (about 300 times slower in my machine!) 

I wanted to test this, I was a bit surprised about the 300 magnitude order. The next Figure shows a comparison between both method when increasing the size of the diagonal (I was a bit bored tonight ...):
![image](https://user-images.githubusercontent.com/26163575/155210521-0f2c157c-3bd6-4a08-ad0e-299a5bddeeac.png)

The Numpy approach is always faster, but I think that even for a diagonal of 300000 points, it takes less than 10ms. I would not choose the Numpy approach over the Scipy looking at the performances. But because you mention that diags skips 0 values, that is not good for us. We need the entire diagonal, otherwise we would shift elements from diagonal, which not desired.

I will go for the Numpy solution, let me know what do you think and update if necessary ;)

Reply:
> Which way should we go to be consistent? Since the collapser bindings use 64-bit, I am tempted to say that we should probably delay conversion to 32-bit to later, so we can use the extra precision if necessary until we really have to downcast inside ripser.

Are you sure about collapser using 64-bit ? Looking at the bindings, the Filtration_value is a simple float (32-bit), in C++ double is 64-bit.

Reply:
> Are you sure about collapser using 64-bit ?
I'm not, no.  I was only fairly confident that GUDHI tend to use 64-bit precision for persistent homology, so expected the same of their collapser and hence also ours.

Reply:
> I'm not, no. I was only fairly confident that GUDHI tend to use 64-bit precision for persistent homology, so expected the same of their collapser and hence also ours.

How, in our case, I use 32-bit because also ripser uses 32-bit, I did not see the point of computing one operation in 32-bit and the other in 64-bit. This shall be addressed in a future PR for support for 64-bit precision ;) (Hopefully soon, if I have some time)

Reply:
I was silly when I talked about 300x speedup, but I do confirm an 80x speedup in the following comparison:
```
diags(dm.diagonal(), format="coo")
```
vs
```
diag_data = dm.diagonal()
row = np.arange(100)
col = row.copy()
```
for dm a 100 x 100 matrix.  The first test takes 40 mus while the second takes 500 ns.

Reply:
I see why. In my opinion with "small" arrays, we maybe pay the price of all the checks done inside of `diags`.
I was just surprised about the difference 😄 .
Thank you for pointing out performances and also the behavior of `diags`, discarding zeros in our case is not the intended behavior.